reference.md

```diff
@@ -499,18 +542,154 @@ variables are missing:
+outer binding with the unary `(= my-symbol)` form. The unary `(= my-symbol)`
+form is *only* valid in a `case` pattern.
+
+Example:
+
+```fennel
+(let [x 1]
+  (case [:hello]
+    ;; 1 != :hello
+    [(= x)] x
+    _ :no-match)) ; -> no-match
+
+(let [x 1]
+  (case [1]
+    ;; 1 == 1
+    [(= x)] x
+    _ :no-match)) ; -> 1
+
+(let [pass :hunter2]
+  (case (user-input)
+    (= pass) :login
+    _ :try-again!))
+```
+
+Unification is only required inside the pattern. Outer bindings are
+automatically available inside guards and bodies as long as the name has not
+been rebound in the pattern.
+
+Example:
+
+```fennel
+(let [limit 10]
+  (case [9 9]
+    ;; automatically avaliable in the guard
+    (where [a b] (< limit (+ a b))) :over-the-limit
+    ;; also available in the body
+    [a b] (+ a b limit))) ; -> :over-the-limit
+```
+
+If a pattern defines a binding with the same name as an existing binding in the
+outer scope, the inner binding takes precedence.
+
+Example:
+
+```fennel
+(let [x 99]
+  (case [10 20]
+    ;; x is not unified, so it's a new binding
+    [x y] (+ x y)) ; -> (+ 10 20)
+
+  (case [10 20]
+    ;; x was not rebound, so the existing symbol is used
+    [a b] (+ a x)) ; -> (+ 10 99)
+
+  (case [99 20]
+    ;; [99 20] = [99 b]
+    [(= x) b] (+ x b)) ; -> (+ 99 20)
+
+  (case [99 20]
+    ;; [99 20] = [99 x]
+    ;; note that x is bound in the pattern, so the body uses that value
+    [(= x) x] (+ x x)) ; -> (+ 20 20)
+
+  (case [10 20]
+    ;; (< (+ 10 20) 99)
+    ;; note that we *must not* rebind x in the pattern if we wish to reference
+    ;; it in a guard or body.
+    (where [p-x y] (< (+ p-x y) x)) (+ p-x y))) ; -> (+ 10 20)
+```
+
+```fennel
+(let [pass :hunter2
+      2fa-key (current-2fa-key)]
+  (case [(input) (2fa)]
+    (where [(= pass) nil]) :login-but-warn-user-they-are-naughty
+    (where [(= pass) 2fa] (valid-2fa? 2fa 2fa-key)) true
+    _ false))
+```
+
+**Note:** The `case` macro can be used in place of the `if-let` macro
+from Clojure. The reason Fennel doesn't have `if-let` is that `case`
+makes it redundant.
+
+### `match` pattern matching
+
+*(since 0.2.0)*
+
+`match` is conceptually equivalent to `case`, except symbols in the patterns are
+always unified with outer-scope symbols if they exist.
+
+It supports all the same syntax as described in `case` except the unification
+(`(= my-symbol)`) expression, as it is always performed.
+
+> Be careful when using `match` that your symbols are not accidentally the same
+> as any existing symbols! If you know you don't intend to unify with any
+> existing symbols you should use the `case` expression.
+
+```fennel
+(let [x 95]
+ (match [52 85 95]
+   [b a a] :no ; because a=85 and a=95
+   [x y z] :no ; because x=95 and x=52
+   [a b x] :yes)) ; a and b are fresh values while x=95 and x=95
+```
+
+Eixisting bindings with the value of `nil` do not follow the `?` prefix rules
+when unified.
```

Comment:
The `?` prefix rule says that locals prefixed with `?` are allowed to match nil values, which is still true if there is an existing binding. Maybe it would be clearer to say that if it is an existing binding, then the `?` prefix is not necessary because the nil-check is bypassed.

Reply:
https://github.com/bakpakin/Fennel/pull/434/commits/bcc389d823f18bbe264e5f2b0778afdf966b13c6

not 100% this is great.